dts/bindings/arm/nordic,nrf21540-fem.yaml

```diff
@@ -0,0 +1,56 @@
+# Copyright (c) 2020 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+
+description: >
+    This is a representation of the nRF21540 Radio Front-End module
+compatible: "nordic,nrf21540-fem"
+
+include: base.yaml
+
+properties:
+    tx-en-gpios:
+        type: phandle-array
+        required: false
+        description: >
+            Gpio of the SOC controlling TX_EN pin of the nRF21540
+    rx-en-gpios:
+        type: phandle-array
+        required: false
+        description: >
+            Gpio of the SOC controlling RX_EN pin of the nRF21540
+    pdn-gpios:
+        type: phandle-array
+        required: false
+        description: >
+            Gpio of the SOC controlling PDN pin of the nRF21540
+    tx-en-settle-time-us:
+        type: int
+        default: 13
```

Comment:
There's concern in issue #25164 about use of binding defaults that might produce incorrect behavior on variant hardware.  Some proposed guidelines are [here](https://github.com/zephyrproject-rtos/zephyr/issues/25164#issuecomment-701486128).

I don't see anything in the data sheet that describes the settling and hold times, which suggests these are very much hardware-dependent.  In that case these should be made required, and the description updated to provide an example value (and perhaps mention how it can be determined).  For convenience a devicetree source fragment showing the default settings can be placed in the description of the binding itself (at the top of the file), where a user can simply copy it into the devicetree source.

If these are conservative settings that would work for any conceivable hardware platform, but might be optimized to shorter durations in certain cases, then they could remain defaults, but the description should explain how they were selected.

If they're "the values that result from using the Nordic reference design" then documenting that could be enough to justify making them defaults.

For values like gain which are presumably configurable (even if the API isn't there yet) these can be defaulted as long as the selected default value is justified in the documentation, e.g. 20 dB is the nominal value of the power-up default.

Reply:
Please have a look at nRF21540 Product Specification: https://infocenter.nordicsemi.com/topic/struct_fem/struct/nrf21540_ps.html table 6. There are defined timings that are used as defaults in this file (I see now that our timings from OPS had different value than from PS, @JakubPegza please have a look and update wherever necessary). These timings are recommended for most designs. However, I'm aware of FEM users that tweak timings in their products. Because of that we should keep them configurable. Do you think that we should add reference to table 6 from Product Specification to the description of these fields?

Values in gain fields are default out-of-factory values. You can find them in the Product Specification as well. It is possible to modify them with SPI interface. If default values are modified with SPI interface, DTS should be modified to reflect amplification value. Should we reference to Product Specification to justify these defaults as well?

Reply:
Yes, if the binding documentation includes most of the explanatory material above and references the specification that should cover it.  At least for the timings.

I would expect a SPI control driver to use the devicetree gain values to reconfigure the FEM automatically when it's initialized.  In that case I would remove the default, and leave the properties as `required: false`, then document that absence of the property means the power-up default will be used.  A compile-time diagnostic would be appropriate if the property is present but there's no way to set it.

Looking at the specification more closely I might use with properties like this:
```
    tx-gain = <22>; /* integer value for CONFREG0.TX_GAIN */
    tx-gain-poutb;   /* boolean flag selecting CONFREG0.MODE = 1 */
    pouta-sel-uicr;  /* boolean flag selecting CONFREG2.POUTA_SEL = 1 */
```
that more closely track the real interface.  If you specify gains as dB, will the driver be able to automatically translate that into a correct `TX_GAIN` register setting?

But that level of detail depends on what your users will expect.  From the Zephyr perspective we just need to know what gets a default, where it came from, and conditions under which it needs to be changed.

Reply:
I'm thinking about this TX and RX gain values. Basically there are multiple values that could be added to DTS binding:

1. POUTA_PROD (constant = 20)
2. POUTA_UICR (optional, can be configured by board creator)
3. POUTA_SEL (bool, if PROD or UICR is being used)

and the same set for POUTB registers.

I'm not sure if there is any point in defining immutable constants like POUTA_PROD in device tree, or device tree bindings. It could be provided as a define in source file. `POUTX_UICR` and `POUTX_SEL` could be added as optional fields to DTS, where board designer could state what values are programmed in nRF21540 UICR for given board. Software could use these values to determine if MODE pin should be asserted or deasserted for given transmission.

However, I think right now power mapping, or power control modules are not implemented and MODE pin is not managed by nRF21540 driver. Because of that I think that we could remove gain-related fields from this PR and return to this discussion when these modules are implemented and could use data from these fields. @pabigot , @JakubPegza , @AdamZelikNS, what do you think?

Reply:
The devicetree model isn't designed for named constants.  But yes, the short path is to remove the stuff you aren't ready to support and add it when there's something that would use it.

Reply:
I agree that if we want to add these fields incrementally, gain configuration is not necessary now. But if only `POUTX_UICR` was to be added (ant not `POUTX_PROD`), what's the use of adding `POUTX_SEL`? Selection could be done by the presence or absence of `POUTX_PROD` field then.

Reply:
@pabigot , have I understand correctly that if we add references to nRF21540 Product Specification to timing `description` fields in the binding file and remove gain fields, you'll approve? Or do you expect any more modifications to this PR?

@JakubPegza , below a little longer explanation why I think that _SEL are needed in DTS to implement power control module.

> But if only `POUTX_UICR` was to be added (ant not `POUTX_PROD`), what's the use of adding `POUTX_SEL`?

TX power can be modified by firmware in two ways:

1. SPI control: using TX_GAIN field in CONFREG0
2. MODE pin control (when SPI is not connected from MCU do FEM)

In SPI mode it is easy: software selects any gain it wants.

But in GPIO mode it is more complicated. Software can select gain of RADIO peripheral and one of two values of gain in nRF21540. Default factory values of nRF21540 gain are POUTA_PROD (20) and POUTB_PROD(10). But these values may be replaced in factory programming UICR of nRF21540. To replace POUTA_PROD, you can program POUTA_UICR with selected value and set POUTA_SEL to 1 to start using POUTA_UICR instead of POUTB_UICR. Similarly POUTB may be replaced.

The only action firmware can do is to set MODE to 0 or to 1. When MODE is set to 0 then POUTA is used, when 1 POUTB is used. But firmware must know what gain is programmed in UICR for POUTA and POUTB to correctly select. Because of that firmware should know POUTx_UICR and POUTx_SEL to decide how MODE pin should be managed. And I think that these values could be added to DTS in the future as optional values.

Reply:
> if we add references to nRF21540 Product Specification ... you'll approve? 

Yes.  Something like:

    Default value is from Table 6 of the nRF21540 Product Specification (v1.0), 
    and can be overridden for tuned configurations.

Alternatively just remove them entirely, store the default values in the driver, and provide API in the driver to update the values under application control.  That would be the most flexible solution: maybe people will care about sub-microsecond differences, and perhaps changing based on operational state, so the user API could use `ns` values.